pkg/database/dbaas.go

```diff
@@ -0,0 +1,104 @@
+package database
+
+import (
+	"fmt"
+	"strings"
+
+	pipeline "github.com/ccremer/go-command-pipeline"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// DBaaSDatabase contains the Database struct needed with the plan, specific of DBaaS
+type DBaaSDatabase struct {
+	Database
+	plan *string
+}
+
+// Execute starts the saving process of the data in the billing database
+func (d *DBaaSDatabase) Execute(dctx *Context) error {
+	p := pipeline.NewPipeline[*Context]()
+	p.WithSteps(
+		p.NewStep("Open database connection", d.openConnection),
+		p.WithNestedSteps("Save initial billing configuration", nil,
+			p.NewStep("Begin transaction", d.beginTransaction),
+			p.NewStep("Ensure initial billing database configuration", d.ensureInitConfiguration),
+			p.NewStep("Commit transaction", d.commitTransaction),
+		).WithErrorHandler(d.rollback),
+		p.NewStep("Save DBaaS usage to billing database", d.saveUsageToDatabase),
+		p.NewStep("Close database connection", d.closeConnection),
+	)
+	return p.RunWithContext(dctx)
+}
+
+// saveUsageToDatabase saves each previously aggregated DBaaS to the billing database
+func (d *DBaaSDatabase) saveUsageToDatabase(dctx *Context) error {
+	log := ctrl.LoggerFrom(dctx)
+	for _, aggregated := range *dctx.AggregatedObjects {
+		err := d.ensureAggregatedServiceUsage(dctx, aggregated)
+		if err != nil {
+			log.Error(err, "Cannot save aggregated DBaaS service record to billing database")
```

Comment:
should we add some log info about which part failed or does `err` have that already?

Reply:
the error is supposed to have the details what failed in the pipeline. So we might need to check in the underlying pipeline if we catch every relevant error.

Reply:
the error has it already. So for example, if the app fails in step `Save facts`, the log will look like:
```
...
ERROR | exoscale-metrics-collector.dbaas | database/dbaas.go:39 Cannot save aggregated DBaaS service record to billing database | {"error": "step 'Saving DBaaS usage namespace default' failed: error found in pipeline: step 'Save facts' failed: some error"}
...
```